addons/web/static/tests/views/form/form_renderer.test.js

```diff
@@ -0,0 +1,127 @@
+import { expect, test } from "@odoo/hoot";
+import { queryAllTexts } from "@odoo/hoot-dom";
+import { Component, xml } from "@odoo/owl";
+import {
+    defineModels,
+    fields,
+    models,
+    mountView,
+    registerMock,
+} from "@web/../tests/web_test_helpers";
+
+class Partner extends models.Model {
+    name = fields.Char();
+    charfield = fields.Char();
+
+    _records = [{ id: 1, name: "firstRecord", charfield: "content of charfield" }];
+}
+defineModels([Partner]);
+
+test("compile form with modifiers", async () => {
+    Partner._views = {
+        form: /*xml*/ `
+            <form>
+                <div invisible="display_name == uid">
+                    <field name="charfield"/>
+                </div>
+                <field name="display_name" readonly="display_name == uid"/>
+            </form>
+        `,
+    };
+
+    await mountView({
+        resModel: "partner",
+        type: "form",
+        resId: 1,
+    });
+    expect(`.o_form_editable input`).toHaveCount(2);
+});
+
+test("compile notebook with modifiers", async () => {
+    Partner._views = {
+        form: /*xml*/ `
+            <form>
+                <sheet>
+                    <notebook>
+                        <page name="p1" invisible="display_name == 'lol'"><field name="charfield"/></page>
+                        <page name="p2"><field name="display_name"/></page>
+                    </notebook>
+                </sheet>
+            </form>
+        `,
+    };
+
+    await mountView({
+        resModel: "partner",
+        type: "form",
+        resId: 1,
+    });
+    expect(queryAllTexts`.o_notebook_headers .nav-item`).toEqual(["p1", "p2"]);
+});
+
+test("compile header and buttons", async () => {
+    Partner._views = {
+        form: /*xml*/ `
+            <form>
+                <header>
+                    <button string="ActionButton" class="oe_highlight" name="action_button" type="object"/>
+                </header>
+            </form>
+        `,
+    };
+
+    await mountView({
+        resModel: "partner",
+        type: "form",
+        resId: 1,
+    });
+    expect(`.o_statusbar_buttons button[name=action_button]:contains(ActionButton)`).toHaveCount(1);
+});
+
+test("render field with placeholder", async () => {
+    registerMock(["fields", "char"], {
```

Comment:
```suggestion
    patchRegistry(registry.category("fields"), "char", { ... });
```
arguments:
 - registerMock sounds a bit weird
 - I'm not a fan of the "path" API. Typically, you'll have a direct reference to the registry you're interested in anyway (fieldRegistry, debugRegistry...)

My first thoughts were that we don't really need an helper for this, but as hoot acts differently than qunit regarding the cleanup, this might be useful indeed.

Reply:
Maybe @Arcasias would like to comment on this as well

Reply:
The purpose of `registerMock` is to not import both `registry` and a helper function that will patch the registry. Just import the registry patch function.

I added the helper because I had to write this big block.
```js
    const fieldsRegistry = registry.category("fields");
    const oldCharField = fieldsRegistry.get("char");
    const charField = {
        component: class CharField extends Component {
            static props = ["*"];
            static template = xml`<div/>`;
            setup() {
                expect.step("setup field component");
                expect(this.props.placeholder).toBe("e.g. Contact's Name or //someinfo...");
            }
        },
        extractProps: ({ attrs }) => ({ placeholder: attrs.placeholder }),
    };
    fieldsRegistry.add("char", charField, { force: true });
    after(() => {
        fieldsRegistry.add("char", oldCharField, { force: true });
    });
```

Reply:
Yep I understand the need. But importing registry is no big deal imo, we import it almost everywhere anyway.